be/src/types/bitmap_value.cpp

```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
+            abs_offset = cardinality + offset;
+        }
+
+        int64_t count = 0;
+        int64_t offset_count = 0;
+        auto it = _bitmap->begin();
+        for (; it != _bitmap->end() && offset_count < abs_offset; ++it) {
+            ++offset_count;
+        }
+        for (; it != _bitmap->end() && count < len; ++it, ++count) {
+            ret_bitmap->add(*it);
+        }
+        return count;
     }
-    return count;
 }
 
 int64_t BitmapValue::bitmap_subset_limit_internal(const int64_t& range_start, const int64_t& limit,
-
                                                   BitmapValue* ret_bitmap) {
-    bool is_reverse = false;
-
-    int64_t abs_limit = limit;
-    if (limit < 0) {
-        is_reverse = true;
-        abs_limit = -limit;
-    }
-
-    int64_t count = 0;
-
-    if (is_reverse) {
-        detail::Roaring64Map::const_iterator start = _bitmap->begin();
-        detail::Roaring64Map::const_iterator end = _bitmap->begin();
-
-        int64_t offset = 0;
-        for (; end != _bitmap->end() && offset < abs_limit && *end <= range_start;) {
-            ++end;
-            ++offset;
+    switch (_type) {
+    case EMPTY:
+        return 0;
+    case SINGLE: {
+        if (_sv < range_start) {
+            return 0;
+        } else {
+            ret_bitmap->add(_sv);
+            return 1;
         }
-        if (offset == abs_limit) {
-            for (; end != _bitmap->end() && *end <= range_start;) {
-                ++start;
+    }
+    case SET: {
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto start = values.begin();
+            auto end = values.begin();
+
+            int64_t offset = 0;
+            for (; end != values.end() && offset < abs_limit && *end <= range_start;) {
                 ++end;
+                ++offset;
+            }
+            if (offset == abs_limit) {
+                for (; end != values.end() && *end <= range_start;) {
+                    ++start;
+                    ++end;
+                }
+            }
+            for (; start != end; ++start, ++count) {
+                ret_bitmap->add(*start);
+            }
+        } else {
+            auto it = values.begin();
+            for (; it != values.end() && *it < range_start;) {
+                ++it;
+            }
+            for (; it != values.end() && count < limit; ++it, ++count) {
+                ret_bitmap->add(*it);
             }
         }
+        return count;
+    }
+    default:
+        DCHECK_EQ(_type, BITMAP);
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto start = _bitmap->begin();
+            auto end = _bitmap->begin();
+
+            int64_t offset = 0;
+            for (; end != _bitmap->end() && offset < abs_limit && *end <= range_start;) {
```

Comment:
/**
     * Returns the number of integers that are smaller or equal to x.
     */
    uint64_t rank(uint64_t x) const {

Reply:
To use this, `Roaring64MapSetBitForwardIterator` need support `move`  op, can I optimize this later?